File: backend/app/models/applications.py
from __future__ import annotations

import uuid

from sqlalchemy import DateTime, ForeignKey, Index, Integer, String, Text
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, TimestampMixin, UUIDPrimaryKeyMixin

APPLICATION_STATUSES = (
    "draft",
    "in_progress",
    "submitted",
    "under_review",
    "documents_pending",
    "approved",
    "rejected",
    "withdrawn",
)


class Application(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "applications"
    __table_args__ = (
        # Serves analytics date-range queries (Task 013).
        Index("ix_applications_college_created", "college_id", "created_at"),
    )

    college_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("colleges.id"), nullable=False, index=True
    )
    student_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("students.id"), nullable=False, index=True
    )
    course_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("courses.id"), nullable=False, index=True
    )
    application_number: Mapped[str | None] = mapped_column(String(100), nullable=True)
    intake: Mapped[str | None] = mapped_column(String(20), nullable=True)
    status: Mapped[str] = mapped_column(String(50), nullable=False, default="draft", index=True)
    completion_percentage: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    submitted_at: Mapped[DateTime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    reviewed_at: Mapped[DateTime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    idempotency_key: Mapped[str | None] = mapped_column(String(255), nullable=True, index=True)


class ApplicationDocument(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "application_documents"

    college_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("colleges.id"), nullable=False, index=True
    )
    application_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("applications.id"), nullable=False, index=True
    )
    document_type: Mapped[str] = mapped_column(String(100), nullable=False)
    file_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    file_url: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(String(50), nullable=False, default="pending")
    verified_at: Mapped[DateTime | None] = mapped_column(DateTime(timezone=True), nullable=True)

File: frontend/types/applications.ts
export type ApplicationStatus =
  | "draft"
  | "in_progress"
  | "submitted"
  | "under_review"
  | "documents_pending"
  | "approved"
  | "rejected"
  | "withdrawn";

export type Application = {
  id: string;
  college_id: string;
  student_id: string;
  student_name: string | null;
  course_id: string;
  course_name: string | null;
  application_number: string | null;
  intake: string | null;
  status: ApplicationStatus | string;
  completion_percentage: number;
  notes: string | null;
  submitted_at: string | null;
  reviewed_at: string | null;
  created_at: string;
  updated_at: string;
};

export type ApplicationListParams = {
  status?: string;
  course_id?: string;
  student_id?: string;
  page?: number;
  page_size?: number;
};

export type ApplicationStatusReport = {
  application_id: string;
  status: string;
  completion_percentage: number;
  missing_information: string[];
  next_steps: string[];
  application_number: string | null;
  submitted_at: string | null;
};

export type ApplicationDocumentStatus = "missing" | "pending" | "verified" | "rejected" | string;

export type ApplicationChecklistItem = {
  document_type: string;
  mandatory?: boolean;
  status: ApplicationDocumentStatus;
  document_id: string | null;
};

export type ApplicationDocument = {
  id: string;
  application_id: string;
  document_type: string;
  file_name: string | null;
  file_url: string | null;
  status: ApplicationDocumentStatus;
  verified_at: string | null;
  created_at: string;
};
